README.md
# ![Logo](https://foobar-land.github.io/coq-of-ocaml/img/rooster-48.png) coq-of-ocaml [![CI](https://github.com/clarus/coq-of-ocaml/workflows/CI/badge.svg?branch=master)](https://github.com/clarus/coq-of-ocaml/actions?query=workflow%3ACI)
> Import OCaml programs to Coq for formal verification

**Documentation on https://foobar-land.github.io/coq-of-ocaml/**

## Aim
`coq-of-ocaml` aims to enable formal verification of [OCaml](https://ocaml.org/) programs&nbsp;🦄. *The more you prove, the happier you are.* By transforming OCaml code into similar [Coq](https://coq.inria.fr/) programs, it is possible to prove arbitrarily complex properties using the existing power of Coq. The sweet spot of `coq-of-ocaml` is purely functional and monadic programs. Side-effects outside of a monad, like references, and advanced features like object-oriented programming, may never be supported. By sticking to the supported subset of OCaml, you should be able to import millions of lines of code to Coq and write proofs at large. Running `coq-of-ocaml` after each code change, you can make sure that your proofs are still valid. The generated Coq code is designed to be stable, with no generated variable names for example. We recommend organizing your proof files as you would organize your unit-test files.

The guiding idea of `coq-of-ocaml` is [TypeScript](https://www.typescriptlang.org/). Instead of bringing types to an untyped language, we are bringing proofs to an already typed language. The approach stays the same: finding the right sweet spot, using heuristics when needed, guiding the user with error messages. We use `coq-of-ocaml` at [Tezos](https://tezos.com/), a crypto-currency implemented in OCaml, in the hope to have near-zero bugs thanks to formal proofs. Tezos is currently one of the most advanced crypto-currencies, with smart contracts, proof-of-stake, encrypted transactions, and protocol upgrades. It aims to compete with Ethereum. Formal verification is claimed to be important for crypto-currencies as there are no central authorities to forbid bug exploits and a lot of money at stake. A Coq translation of the core components of Tezos is available in the project [coq-tezos-of-ocaml](https://gitlab.com/nomadic-labs/coq-tezos-of-ocaml). Protecting the money.

There are still some open problems with `coq-of-ocaml`, like the axiom-free compilation of [GADTs](https://blog.janestreet.com/why-gadts-matter-for-performance/) (an ongoing project). If you are willing to work on a particular project, please contact us by opening an issue in this repository.

<p align="center">
  <img alt="happiness and proofs" width="347" height="auto" src="https://raw.githubusercontent.com/clarus/coq-of-ocaml/master/doc/proofs_happiness.png" />
</p>

## Example
Start with the file `main.ml`&nbsp;🐫:
```ocaml
type 'a tree =
  | Leaf of 'a
  | Node of 'a tree * 'a tree

let rec sum tree =
  match tree with
  | Leaf n -> n
  | Node (tree1, tree2) -> sum tree1 + sum tree2
```
Run:
```
coq-of-ocaml main.ml
```
Get a file `Main.v`&nbsp;🦄:
```coq
Require Import CoqOfOCaml.CoqOfOCaml.
Require Import CoqOfOCaml.Settings.

Inductive tree (a : Set) : Set :=
| Leaf : a -> tree a
| Node : tree a -> tree a -> tree a.

Arguments Leaf {_}.
Arguments Node {_}.

Fixpoint sum (tree : tree int) : int :=
  match tree with
  | Leaf n => n
  | Node tree1 tree2 => Z.add (sum tree1) (sum tree2)
  end.
```
You can now write proofs by induction over the `sum` function using Coq. To see how you can write proofs, you can simply look at the [Coq documentation](https://coq.inria.fr/documentation). Learning to write proofs is like learning a new programming paradigm. It can take time, but be worthwhile! Here is an example of proof:
```coq
(** Definition of a tree with only positive integers *)
Inductive positive : tree int -> Prop :=
| Positive_leaf : forall n, n > 0 -> positive (Leaf n)
| Positive_node : forall tree1 tree2,
  positive tree1 -> positive tree2 -> positive (Node tree1 tree2).

Require Import Coq.micromega.Lia.

Lemma positive_plus n m : n > 0 -> m > 0 -> n + m > 0.
  lia.
Qed.

(** Proof that if a tree is positive, then its sum is positive too *)
Fixpoint positive_sum (tree : tree int) (H : positive tree)
  : sum tree > 0.
  destruct tree; simpl; inversion H; trivial.
  apply positive_plus; now apply positive_sum.
Qed.
```

## Install
Using the OCaml package manager [opam](https://opam.ocaml.org/), run:
```
opam install coq-of-ocaml
```

## Usage
The basic command is:
```
coq-of-ocaml file.ml
```
You can start to experiment with the test files in `tests/` or look at our [online examples](https://foobar-land.github.io/coq-of-ocaml/examples/). `coq-of-ocaml` compiles the `.ml` or `.mli` files using [Merlin](https://github.com/ocaml/merlin) to understand the dependencies of a project. One first needs to have a **compiled project** with a working configuration of Merlin. This is automatically the case if you use [dune](https://dune.build/) as a build system.

## Documentation
You can read the documentation on the website of the project at [https://foobar-land.github.io/coq-of-ocaml/](https://foobar-land.github.io/coq-of-ocaml/).

## Supported
* the core of OCaml (functions, let bindings, pattern-matching,...) ✔️
* type definitions (records, inductive types, synonyms, mutual types) ✔️
* monadic programs ✔️
* modules as namespaces ✔️
* modules as polymorphic records (signatures, functors, first-class modules) ✔️
* multiple-file projects (thanks to Merlin) ✔️
* `.ml` and `.mli` files ✔️
* existential types (we use impredicative sets to avoid a universe explosion) ✔️
* partial support of GADTs 🌊
* partial support of polymorphic variants 🌊
* partial support of extensible types 🌊
* ignores side-effects outside of a monad ❌
* no object-oriented programming ❌

Even in case of errors, we try to generate some Coq code along with an error message. The generated Coq code should be readable and with a size similar to the OCaml source. The generated code does not necessarily compile after a first try. This can be due to various errors, such as name collisions. Do not hesitate to fix these errors by updating the OCaml source accordingly. If you want more assistance, please contact us by opening an issue in this repository.

## Contribute
If you want to contribute to the project, you can submit a pull-requests.

### Build with opam
To install the current development version:
```
opam pin add https://github.com/clarus/coq-of-ocaml.git#master
```

### Build manually
Clone the Git submodule for [Merlin](https://github.com/ocaml/merlin):
```
git submodule init
git submodule update
```
Then read the `coq-of-ocaml.opam` file at the root of the project to know the dependencies to install and get the list of commands to build the project.

## License
MIT (open-source software)


CHANGELOG.md
## 2.5.2 (January 25, 2022)
* Various small improvements.

## 2.5.1 (June 24, 2021)
* Improvement of the GADTs support.

## 2.5.0 (March 30, 2021)
* Beginning of support of GADTs without axioms, thanks to a pull-request of @pedrotst https://github.com/clarus/coq-of-ocaml/pull/166.
* Upgrade OCaml to the version 4.12, thanks to @lthms.

## 2.4.1 (March 15, 2021)
* Add basic translation of `try ... with` with extensible types (cannot run in Coq but may be extracted to OCaml).
* Add basic support for matching on extensible types.
* Add the attribute `@coq_type_annotation` to generate the type annotation of an expression.
* Show nicer error messages for errors in the configuration file.
* Remove existential types from the modules (except for the first-class modules).
* Upgrade OCaml to the version 4.10.
* Upgrade Dune to the version 2.8.

## 2.4.0 (January 11, 2021)
* Install the Coq proofs in `CoqOfOCaml` rather than `OCaml` for clarity.
* Simpler header at the top of the generated files.
* Try to import documentation comments from OCaml.
* Do not ignore sequences of instructions anymore.
* Add support for top-level definitions as `init_module` functions.
* Add the attribute `@coq_mutual_as_notation` to translate a mutually recursive function as a notation.
* Add the configuration option `renaming_type_constructor` to rename some type constructors.
* Add the configuration option `operator_infix` to have notations for the infix operators.
* Add the configuration option `constant_warning` to disable warnings when converting constants.
* Add the configuration option `first_class_module_signature_blacklist` to ignore some signatures in the search of a signature name.

## 2.3.0 (November 3, 2020)
* Add support for the OCaml's monadic notation.
* Add support for records with polymorphic fields.
* Add the attribute `@coq_cast` to force an unsafe cast of a sub-expression.
* Rename the attribute `@coq_axiom` as `@coq_axiom_with_reason`.
* Add the attribute `@coq_precise_signature` to help to distinguish between ambiguous signatures.
* Remove the generation of `Import` in Coq and expand the names regardless of the `open` commands, in order to be compatible with local opens.
* Add support for anonymous sub-signatures.
* Generate functors using a type class of the current arguments.
* Add the configuration option `merge_returns`.
* Add the configuration option `merge_types`.
* Add the configuration options `monadic_lets` (renaming of `monadic_bind`), `monadic_let_returns`, `monadic_rets` and `monadic_ret_lets`.
* Add a `@coq_plain_module` attribute to force a module to be compiled as a plain module.
* Add the configuration option `renaming_rules`.
* Add support for generative functors (using functors with a `unit` parameter).
* Add a mandatory string parameter to the `@coq_axiom` attribute in order to give a reason for the axiom.

## 2.2.1 (June 19, 2020)
* Use a more standard Dune build command by keeping in the source the parser generated by [Menhir](https://gitlab.inria.fr/fpottier/menhir).

## 2.2.0 (June 17, 2020)
* Ignore unreachable expressions in match cases.
* Add configuration to black list errors containing a certain message.
* Add configuration to disable exit with error code on some files.
* Exit with a non-null code in case of error.
* Add configuration for the requires from long ident.
* Add configuration to find the type of a polymorphic variant.
* Add configuration to blacklist some errors.
* Add configuration to blacklist some first-class module paths.
* Add configuration to rename some variants.
* Add configuration to rename some constructors.
* Add configuration for module barriers in record aliases.
* Add configuration to escape some value names.
* Add configuration to add head suffix in the generated file.
* Add configuration to add monadic operators.
* Add configuration to specify dependencies required as mli files.
* Add configuration to list the required files in a file, together with an import or not.
* Add configuration to disable guard or positivity checking.
* Add handling of a configuration file.
* Add the `@coq_phantom` attribute to force an abstract type declaration to be phantom.

## 2.1.0 (March 20, 2020)
* Add automatic re-ordering of type synonyms in mutual types to generate valid definitions.
* Add minimal handling of class types as records.
* Ignore type parameters with constraints (like being a sub-type of some variants).
* Add a `@coq_struct "param"` attribute to specify the decreasing parameter name of fixpoints.
* Add a tactic `rewrite_cast_exists_eval_eq` to simplify the use of the `cast_exists` axiom in proofs.
* Name the arguments of the signatures.
* Rename `obj_magic` as `cast`.
* Set the primitive projection flag.
* Add support for the `with` operator on constructor records.
* Add an attribute `@coq_match_gadt_with_result` for GADT matches with casts for the results.
* Do not generate casts for the return values of the match branches with `@coq_match_gadt`.
* Remove the rarely used `match exception when false` construct for default return value in matches.
* Add arity annotations for the existentials.
* Eliminate phantom types and propagate this erasing.
* Inline the application operators `@@` and `|>`.
* Put first-class modules in `Set`, using existentials in impredicative sets.
* Add a `@coq_match_with_default` to generate a default branch for incomplete matches.
* Add a `@coq_force_gadt` attribute to force a type to be defined as a GADT (without type parameters).
* Add basic handling of module alias and typeof in `.mli` files.
* Add more type annotations on values to better support polymorphic values.
* Add better support for include of signatures in `.mli` files.
* Add support of `open` in `.mli` files.
* Upgrade to Coq 8.11.
* Add support for functors inside signatures.
* Add a special treatment with no axioms for the matching of algebraic types with existentials which are not GADTs (with the same type parameters for all the constructors).

make.rb
# encoding: UTF-8
require 'erb'
require 'json'
require 'pathname'
include(ERB::Util)

# Command line arguments.
kernel_directory, tezos_directory, tezos_interface_directory = ARGV

def mark_text(text, errors)
  bytes_errors = text.bytes.to_a.map {|byte| {byte: byte, errors: []}}
  global_errors = []
  errors.each do |error|
    if error["location"]["start"] >= 0 && error["location"]["end"] >= 0 then
      error["location"]["start"].upto(error["location"]["end"] - 1) do |byte_index|
        if bytes_errors[byte_index] then
          bytes_errors[byte_index][:errors] << error
        end
      end
    else
      global_errors << error["message"]
    end
  end

  fragments = [{errors: [], text_bytes: []}]
  bytes_errors.each do |current_errors|
    if current_errors[:errors] == fragments[-1][:errors] then
      fragments[-1][:text_bytes] << current_errors[:byte]
    else
      fragments << {errors: current_errors[:errors].dup, text_bytes: [current_errors[:byte]]}
    end
  end

  output = ""
  fragments.each do |fragment|
    encoded_text = h(fragment[:text_bytes].pack("U*"))
    encoded_errors = h(fragment[:errors].map {|error| error["message"]}.join("\n\n" + "—" * 10 + "\n\n"))
    if fragment[:errors].size == 0 then
      output << encoded_text
    else
      output << "<abbr class=\"mark-error\" title=\"#{encoded_errors}\">#{encoded_text}</abbr>"
    end
  end

  [global_errors, output]
end

def get_conversions(directory)
  conversions = []
  ocaml_file_names = Dir.glob(File.join(directory, "**/*.ml{,i}")).to_a.sort.select {|file_name|
    not file_name.include?("test")
  }
  ocaml_file_names.each_with_index do |ocaml_file_name, index|
    print "\r#{directory} (#{index + 1}/#{ocaml_file_names.size})"
    ocaml_name = Pathname.new(ocaml_file_name).relative_path_from(Pathname.new(directory)).to_s
    ocaml_content = File.read(ocaml_file_name, :encoding => 'utf-8')
    coq_extension = File.extname(ocaml_file_name) == ".ml" ? ".v" : "_mli.v"
    ocaml_base_name = File.basename(ocaml_file_name, ".*")
    capitalized_base_name = ocaml_base_name[0].capitalize + ocaml_base_name[1..-1]
    coq_file_name = File.join(
      File.dirname(ocaml_file_name),
      capitalized_base_name + coq_extension
    )
    coq_name = File.basename(coq_file_name)
    errors_file_name = ocaml_file_name + ".errors"
    if File.exists?(coq_file_name) then
      errors_content = File.read(errors_file_name)
      errors_json = errors_content != "" ? JSON.parse(errors_content) : []
      global_errors, marked_ocaml_content = mark_text(ocaml_content, errors_json)
      nb_errors = errors_json.size - global_errors.size
      coq_content = File.read(coq_file_name, :encoding => 'utf-8')
      if coq_content.valid_encoding? then
        conversions << {
          ocaml_name: ocaml_name,
          nb_errors: nb_errors,
          global_errors: [],
          ocaml_content: marked_ocaml_content,
          ocaml_size: ocaml_content.split("\n").size,
          raw_ocaml_content: ocaml_content,
          coq_name: coq_name,
          coq_content: coq_content,
          coq_size: coq_content.split("\n").size
        }
      end
    else
      puts
      puts "'#{coq_file_name}' not found"
    end
  end
  puts
  conversions.sort_by {|conversion| conversion[:ocaml_name]}
end

# Helpers.
def header(root, section)
  ERB.new(File.read("template/header.html.erb")).result(binding)
end

def footer(root)
  ERB.new(File.read("template/footer.html.erb")).result(binding)
end

def project(name, title, intro, block_quote, directory, status, chart_data)
  project_name = name
  project_intro = <<-END
    <h2>
      #{title} in&nbsp;Coq
      <small>
      #{
        case status
        when :does_compile
          "<span class=\"label label-success\">Does compile</span>"
        when :active_development
          "<span class=\"label label-warning\">Active development</span>"
        when :does_not_compile
          "<span class=\"label label-danger\">Does not compile</span>"
        end
      }
      </small>
    </h2>
    <p>#{intro}</p>
    #{"<blockquote class=\"blockquote\"><p class=\"mb-0\" style=\"font-size: 16px;\">#{block_quote}</p></blockquote>" if block_quote}
  END
  conversions = get_conversions(directory)
  nb_ocaml_lines = conversions.reduce(0) {|sum, conversion|
    sum + conversion[:ocaml_size]
  }
  nb_coq_lines = conversions.reduce(0) {|sum, conversion|
    sum + conversion[:coq_size]
  }
  nb_errors = conversions.reduce(0) {|sum, conversion|
    sum + conversion[:nb_errors]
  }
  ERB.new(File.read("template/project.html.erb")).result(binding)
end

# Generate the files.
File.open("index.html", "w") do |file|
  file << ERB.new(File.read("index.html.erb")).result(binding)
end

File.open("kernel/index.html", "w") do |file|
  file << project(
    :kernel,
    "Kernel of Coq",
    "This is a demo of the current development version of <a href=\"https://github.com/clarus/coq-of-ocaml\">coq-of-ocaml</a> on the <a href=\"https://github.com/coq/coq/tree/master/kernel\">kernel</a> of <a =href=\"https://coq.inria.fr/\">Coq</a>. Coq is written in <a =href=\"https://ocaml.org/\">OCaml</a>.",
    nil,
    kernel_directory,
    :does_not_compile,
    nil
  )
end

File.open("tezos/index.html", "w") do |file|
  file << project(
    :tezos,
    "Protocol of Tezos",
    "These are the sources of the <a href=\"https://gitlab.com/tezos/tezos/tree/master/src/proto_alpha/lib_protocol\">protocol</a> of <a href=\"https://tezos.com/\">Tezos</a> imported to <a href=\"https://coq.inria.fr/\">Coq</a> by the current development version of <a href=\"https://github.com/clarus/coq-of-ocaml\">coq-of-ocaml</a>. Tezos is a crypto-currency with smart-contracts and an upgradable protocol.",
    "(2020-01-25) The import of Tezos code to Coq is currently on pause to consolitate gains (work on small features, write documentation). Our next targets will probably be the files with a lot of GADTs and the functors of the storage.",
    tezos_directory,
    :active_development,
    <<-END
      {
        compiling: [
          13,
          292,
          1784,
          1529,
          1147,
          1541,
          2680,
          3100,
          2962,
          5892,
          9302,
          9302,
          9302,
        ],
        generated: [
          47563,
          41904,
          43481,
          43404,
          44618,
          45107,
          44768,
          45158,
          46874,
          46800,
          49535,
          49535,
          49535,
        ],
        labels: [
          "01-14",
          "01-15",
          "01-16",
          "01-17",
          "01-18",
          "01-19",
          "01-20",
          "01-21",
          "01-22",
          "01-23",
          "01-24",
          "01-25",
          "01-26",
        ]
      }
    END
  )
end

File.open("tezos-interface/index.html", "w") do |file|
  file << project(
    :tezos_interface,
    "Interface of the protocol of Tezos",
    "These are the sources of the interface of the <a href=\"https://gitlab.com/tezos/tezos/tree/master/src/proto_alpha/lib_protocol\">protocol</a> of <a href=\"https://tezos.com/\">Tezos</a> imported to <a href=\"https://coq.inria.fr/\">Coq</a> by the current development version of <a href=\"https://github.com/clarus/coq-of-ocaml\">coq-of-ocaml</a>. Tezos is a crypto-currency with smart-contracts and an upgradable protocol.",
    nil,
    tezos_interface_directory,
    :does_compile,
    <<-END
      {
        compiling: [
          7,
          826,
          1190,
          2184,
          6443,
        ],
        generated: [
          2185,
          2200,
          2829,
          6265,
          6443,
        ],
        labels: [
          "01-05",
          "01-06",
          "01-07",
          "01-08",
          "01-09",
        ]
      }
    END
  )
end
